### PR TITLE
feat: Introduce `hawkeye-maven-plugin`

### LINKED ISSUE #21: Distribute as a Maven plugin
So, this project is strongly inspired by [license-maven-plugin](https://github.com/mathieucarbou/license-maven-plugin), with an initial motivation to bring it beyond a Maven plugin.

But, with a core library to support basic check, format, and remove logic, we can evaluate building a maven plugin upon these core functions :)

The significant part is that we should populate properties correctly and do the right overriding. Also, the license-maven-plugin supports a few "multi" configs, while I'd prefer to use multiple executions simply. But this should also be verified.

### PR DESCRIPTION
**Pull request by spencercjh**
RESOLVE #21

### CODE REVIEW

**Comment by tisonkun** [comment]
Thanks for your participantion @spencercjh!

This seems in a good direction. You may depend on the `hawkeye-core` module and following `hawkeye-command` and `license-maven-plugin` to wrap a Mojo facade of `LicenseChecker`, `LicenseFormatter` and `LicenseRemover`.

**Comment by spencercjh** [comment]
@tisonkun I have run `./mvnw spotless:apply` but I fount that the package import order is still not as same as yours.

**Comment by tisonkun** [comment]
@spencercjh You're right that I don't add an `ImportOrder` rule here. Let me fix it today.

**Comment by tisonkun** [comment]
@spencercjh You can merge https://github.com/korandoru/hawkeye/pull/89 and rerun the command above.

**Comment by spencercjh** [comment]
> @spencercjh You can merge #89 and rerun the command above.

done

**Comment by tisonkun** [comment]
Cool! You may try to add some tests or at least share some evidences tested locally so that we're sure it's a good starting point for Maven plugin support.

**Comment by spencercjh** [comment]
> Cool! You may try to add some tests or at least share some evidences tested locally so that we're sure it's a good starting point for Maven plugin support.

done

**Comment by spencercjh** [comment]
I will fix wrong tests ASAP.

**Comment by tisonkun** [comment]
@spencercjh We're not in a hurry so just take your time :D

You can verify the tests locally first.

**Comment by spencercjh** [comment]
> @spencercjh We're not in a hurry so just take your time :D
> 
> You can verify the tests locally first.

It works locally then I commit it.😅

File: hawkeye-maven-plugin/src/test/java/io/korandoru/hawkeye/RemoveMojoTest.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import org.junit.jupiter.api.Test;
+
+class RemoveMojoTest {
+
+    RemoveMojo removeMojo;
+    File tempFile;
+
+    @BeforeEach
+    void setUp() throws IOException {
+        tempFile = File.createTempFile("test", ".yaml", new File("src/test/resources/test_remove"));
```
Comment (spencercjh) [inline hawkeye-maven-plugin/src/test/java/io/korandoru/hawkeye/RemoveMojoTest.java:37]: There are no files added by Git in the dir `test_remove` so the dir doesn't exist so the test failed during GH workflows.

Comment (spencercjh) [reply hawkeye-maven-plugin/src/test/java/io/korandoru/hawkeye/RemoveMojoTest.java]: @tisonkun I think it's good to go for CI now.

File: licenserc.toml [hunk 2]
```diff
@@ -20,6 +20,7 @@ excludes = [
     "**/src/test/**/*.txt",
     "**/src/test/**/*.xml",
 
+    "hawkeye-maven-plugin/src/test/java/io/korandoru/hawkeye/package-info.java",
```
Comment (spencercjh) [inline licenserc.toml:23]: Maybe I SHOULD not use the Java package-info file as a test file. Do you mind it?

Comment (tisonkun) [reply licenserc.toml:23]: Why? If you want to use against Java, you can put one outside the sources tree.

**Comment by spencercjh** [comment]
@tisonkun 

image

I found something unexpected on Windows which leads the failure of my tests.

All `\r\n` in `firstLines` were replaced by `\n` as `fileHeader`.

**But there are still `\r\n` in `expectedHeader`.**

I think it's a bug because of missing integration tests on Windows :D.

**Comment by spencercjh** [comment]
> Let me think of it..
> 
> The expected header should be generated from the content or file you provided instead of the tested files. So I may suggest that you force use `\n` instead of `\r\n` in the template (expected header), rather than do some in place ad-hoc fixup.

Do you mean to force the line separator of file `hawkeye-core/src/main/resources/Apache-2.0.txt`? Or force the line separator of the header file configured in the config toml file?

**Comment by spencercjh** [comment]
Now I found that if I don't update `Header.java` when I add the plugin to `hawkeye-command` and run `./mvnw hawkeye:check`, it will fail and all of the files will be considered as missing header.

**Comment by tisonkun** [comment]
> Do you mean to force the line separator

After a closer look I noticed that we use `endOfLine` variable to build the expected header and update. Then I think we can do the same fixup as your previous commit does.

**Comment by tisonkun** [comment]
> Now I found that if I don't update

May you share the console output or screenshot? I'm not sure what's the issue here.

**Comment by spencercjh** [comment]
image

It runs on b7cb7fd88dc93f93b2d86dcd30909e9e2b05bf4a, the previous commit.

image

> May you share the console output or screenshot? I'm not sure what's the issue here.

I just want to tell you the bug. I think there is no other issues besides of the `endOfLine`.

**Review by tisonkun — APPROVED** [review]
Thank you!

It's a bit late now so I may not be quite clear.

Pending to merge tomorrow morning.

**Comment by tisonkun** [comment]
Merging...

I'll prepare a release for delivering this plugin. Maybe you can try to write a few docs for the Maven plugin usage. Either in Usage section under README.md or in the hawkeye-maven-plugin folder should be fine.

**Comment by tisonkun** [comment]
Released as 3.3.0. We may add a skip option to these Mojos so that users can opt-out the plugin execution.

**Pull request merged**